File: src/tools/pmd-cpd/eval-repos/synthetic/python/heavy_duplication.py
"""Python file with heavy duplication - multiple large duplicated blocks."""


def generate_user_report(users: list[dict]) -> str:
    """Generate a formatted report for users."""
    lines = []
    lines.append("=" * 60)
    lines.append("USER REPORT")
    lines.append("=" * 60)
    lines.append("")

    for item in users:
        lines.append(f"ID: {item.get('id', 'N/A')}")
        lines.append(f"Name: {item.get('name', 'Unknown')}")
        lines.append(f"Email: {item.get('email', 'N/A')}")
        lines.append(f"Status: {item.get('status', 'active')}")
        lines.append(f"Created: {item.get('created_at', 'Unknown')}")
        lines.append("-" * 40)

    lines.append("")
    lines.append(f"Total records: {len(users)}")
    lines.append("=" * 60)
    return "\n".join(lines)


def generate_admin_report(admins: list[dict]) -> str:
    """Generate a formatted report for admins - duplicated structure."""
    lines = []
    lines.append("=" * 60)
    lines.append("ADMIN REPORT")
    lines.append("=" * 60)
    lines.append("")

    for item in admins:
        lines.append(f"ID: {item.get('id', 'N/A')}")
        lines.append(f"Name: {item.get('name', 'Unknown')}")
        lines.append(f"Email: {item.get('email', 'N/A')}")
        lines.append(f"Status: {item.get('status', 'active')}")
        lines.append(f"Created: {item.get('created_at', 'Unknown')}")
        lines.append("-" * 40)

    lines.append("")
    lines.append(f"Total records: {len(admins)}")
    lines.append("=" * 60)
    return "\n".join(lines)


def generate_guest_report(guests: list[dict]) -> str:
    """Generate a formatted report for guests - duplicated structure."""
    lines = []
    lines.append("=" * 60)
    lines.append("GUEST REPORT")
    lines.append("=" * 60)
    lines.append("")

    for item in guests:
        lines.append(f"ID: {item.get('id', 'N/A')}")
        lines.append(f"Name: {item.get('name', 'Unknown')}")
        lines.append(f"Email: {item.get('email', 'N/A')}")
        lines.append(f"Status: {item.get('status', 'active')}")
        lines.append(f"Created: {item.get('created_at', 'Unknown')}")
        lines.append("-" * 40)

    lines.append("")
    lines.append(f"Total records: {len(guests)}")
    lines.append("=" * 60)
    return "\n".join(lines)


def validate_user_input(data: dict) -> list[str]:
    """Validate user input data."""
    errors = []
    if not data.get("name"):
        errors.append("Name is required")
    if not data.get("email"):
        errors.append("Email is required")
    if "@" not in data.get("email", ""):
        errors.append("Invalid email format")
    if not data.get("password"):
        errors.append("Password is required")
    if len(data.get("password", "")) < 8:
        errors.append("Password must be at least 8 characters")
    if not data.get("age"):
        errors.append("Age is required")
    if data.get("age", 0) < 18:
        errors.append("Must be at least 18 years old")
    return errors


def validate_admin_input(data: dict) -> list[str]:
    """Validate admin input data - duplicated validation."""
    errors = []
    if not data.get("name"):
        errors.append("Name is required")
    if not data.get("email"):
        errors.append("Email is required")
    if "@" not in data.get("email", ""):
        errors.append("Invalid email format")
    if not data.get("password"):
        errors.append("Password is required")
    if len(data.get("password", "")) < 8:
        errors.append("Password must be at least 8 characters")
    if not data.get("age"):
        errors.append("Age is required")
    if data.get("age", 0) < 18:
        errors.append("Must be at least 18 years old")
    return errors

File: src/tools/pmd-cpd/eval-repos/synthetic/rust/heavy_duplication.rs
//! Rust file with heavy duplication - multiple large duplicated blocks.

/// Report item representation.
pub struct ReportItem {
    pub id: Option<String>,
    pub name: Option<String>,
    pub email: Option<String>,
    pub status: Option<String>,
    pub created_at: Option<String>,
}

/// Generate a formatted report for users.
pub fn generate_user_report(users: &[ReportItem]) -> String {
    let mut lines = Vec::new();
    lines.push("=".repeat(60));
    lines.push("USER REPORT".to_string());
    lines.push("=".repeat(60));
    lines.push(String::new());

    for item in users {
        let id = item.id.as_deref().unwrap_or("N/A");
        let name = item.name.as_deref().unwrap_or("Unknown");
        let email = item.email.as_deref().unwrap_or("N/A");
        let status = item.status.as_deref().unwrap_or("active");
        let created_at = item.created_at.as_deref().unwrap_or("Unknown");
        lines.push(format!("ID: {}", id));
        lines.push(format!("Name: {}", name));
        lines.push(format!("Email: {}", email));
        lines.push(format!("Status: {}", status));
        lines.push(format!("Created: {}", created_at));
        lines.push("-".repeat(40));
    }

    lines.push(String::new());
    lines.push(format!("Total records: {}", users.len()));
    lines.push("=".repeat(60));
    lines.join("\n")
}

/// Generate a formatted report for admins - duplicated structure.
pub fn generate_admin_report(admins: &[ReportItem]) -> String {
    let mut lines = Vec::new();
    lines.push("=".repeat(60));
    lines.push("ADMIN REPORT".to_string());
    lines.push("=".repeat(60));
    lines.push(String::new());

    for item in admins {
        let id = item.id.as_deref().unwrap_or("N/A");
        let name = item.name.as_deref().unwrap_or("Unknown");
        let email = item.email.as_deref().unwrap_or("N/A");
        let status = item.status.as_deref().unwrap_or("active");
        let created_at = item.created_at.as_deref().unwrap_or("Unknown");
        lines.push(format!("ID: {}", id));
        lines.push(format!("Name: {}", name));
        lines.push(format!("Email: {}", email));
        lines.push(format!("Status: {}", status));
        lines.push(format!("Created: {}", created_at));
        lines.push("-".repeat(40));
    }

    lines.push(String::new());
    lines.push(format!("Total records: {}", admins.len()));
    lines.push("=".repeat(60));
    lines.join("\n")
}

/// Generate a formatted report for guests - duplicated structure.
pub fn generate_guest_report(guests: &[ReportItem]) -> String {
    let mut lines = Vec::new();
    lines.push("=".repeat(60));
    lines.push("GUEST REPORT".to_string());
    lines.push("=".repeat(60));
    lines.push(String::new());

    for item in guests {
        let id = item.id.as_deref().unwrap_or("N/A");
        let name = item.name.as_deref().unwrap_or("Unknown");
        let email = item.email.as_deref().unwrap_or("N/A");
        let status = item.status.as_deref().unwrap_or("active");
        let created_at = item.created_at.as_deref().unwrap_or("Unknown");
        lines.push(format!("ID: {}", id));
        lines.push(format!("Name: {}", name));
        lines.push(format!("Email: {}", email));
        lines.push(format!("Status: {}", status));
        lines.push(format!("Created: {}", created_at));
        lines.push("-".repeat(40));
    }

    lines.push(String::new());
    lines.push(format!("Total records: {}", guests.len()));
    lines.push("=".repeat(60));
    lines.join("\n")
}

/// Input data for validation.
pub struct InputData {
    pub name: Option<String>,
    pub email: Option<String>,
    pub password: Option<String>,
    pub age: Option<u32>,
}

/// Validate user input data.
pub fn validate_user_input(data: &InputData) -> Vec<String> {
    let mut errors = Vec::new();
    if data.name.is_none() || data.name.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Name is required".to_string());
    }
    if data.email.is_none() || data.email.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Email is required".to_string());
    }
    if !data.email.as_ref().map(|s| s.contains('@')).unwrap_or(false) {
        errors.push("Invalid email format".to_string());
    }
    if data.password.is_none() || data.password.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Password is required".to_string());
    }
    if data.password.as_ref().map(|s| s.len()).unwrap_or(0) < 8 {
        errors.push("Password must be at least 8 characters".to_string());
    }
    if data.age.is_none() {
        errors.push("Age is required".to_string());
    }
    if data.age.unwrap_or(0) < 18 {
        errors.push("Must be at least 18 years old".to_string());
    }
    errors
}

/// Validate admin input data - duplicated validation.
pub fn validate_admin_input(data: &InputData) -> Vec<String> {
    let mut errors = Vec::new();
    if data.name.is_none() || data.name.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Name is required".to_string());
    }
    if data.email.is_none() || data.email.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Email is required".to_string());
    }
    if !data.email.as_ref().map(|s| s.contains('@')).unwrap_or(false) {
        errors.push("Invalid email format".to_string());
    }
    if data.password.is_none() || data.password.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        errors.push("Password is required".to_string());
    }
    if data.password.as_ref().map(|s| s.len()).unwrap_or(0) < 8 {
        errors.push("Password must be at least 8 characters".to_string());
    }
    if data.age.is_none() {
        errors.push("Age is required".to_string());
    }
    if data.age.unwrap_or(0) < 18 {
        errors.push("Must be at least 18 years old".to_string());
    }
    errors
}
